x-pack/plugin/security/src/test/java/org/elasticsearch/xpack/security/transport/netty4/SecurityNetty4ServerTransportAuthenticationTests.java

```diff
@@ -0,0 +1,441 @@
+            remoteSecurityNetty4ServerTransport,
+            () -> ThreadPool.terminate(threadPool, 10, TimeUnit.SECONDS)
+        );
+    }
+
+    @SuppressWarnings("unchecked")
+    public void testProxyStrategyConnectionClosesWhenAuthenticatorAlwaysFails() throws Exception {
+        // all requests fail authn
+        doAnswer(invocation -> {
+            ((ActionListener<Void>) invocation.getArguments()[1]).onFailure(new ElasticsearchSecurityException("failed authn"));
+            return null;
+        }).when(remoteCrossClusterAccessAuthenticationService).tryAuthenticate(any(Map.class), anyActionListener());
+        Settings localSettings = Settings.builder()
+            .put(onlyRole(DiscoveryNodeRole.REMOTE_CLUSTER_CLIENT_ROLE))
+            .put(RemoteConnectionStrategy.REMOTE_CONNECTION_MODE.getConcreteSettingForNamespace(remoteClusterName).getKey(), "proxy")
+            .put(
+                ProxyConnectionStrategy.PROXY_ADDRESS.getConcreteSettingForNamespace(remoteClusterName).getKey(),
+                remoteTransportService.boundRemoteAccessAddress().publishAddress().toString()
+            )
+            .put(
+                ProxyConnectionStrategy.REMOTE_SOCKET_CONNECTIONS.getConcreteSettingForNamespace(remoteClusterName).getKey(),
+                randomIntBetween(1, 3) // easier to debug with just 1 connection
+            )
+            .build();
+        {
+            final MockSecureSettings secureSettings = new MockSecureSettings();
+            secureSettings.setString(
+                RemoteClusterService.REMOTE_CLUSTER_CREDENTIALS.getConcreteSettingForNamespace(remoteClusterName).getKey(),
+                randomAlphaOfLength(20)
+            );
+            localSettings = Settings.builder().put(localSettings).setSecureSettings(secureSettings).build();
+        }
+        try (
+            MockTransportService localService = MockTransportService.createNewService(
+                localSettings,
+                VersionInformation.CURRENT,
+                TransportVersion.current(),
+                threadPool
+            )
+        ) {
+            localService.start();
+            RemoteClusterService remoteClusterService = localService.getRemoteClusterService();
+            // obtain some connections and check that they'll be promptly closed
+            for (int i = 0; i < randomIntBetween(4, 16); i++) {
+                PlainActionFuture<Void> closeFuture = PlainActionFuture.newFuture();
+                CountDownLatch connectionTestDone = new CountDownLatch(1);
+                remoteClusterService.maybeEnsureConnectedAndGetConnection(remoteClusterName, true, ActionListener.wrap(connection -> {
+                    // {@code RemoteClusterService.REMOTE_CLUSTER_HANDSHAKE_ACTION_NAME} fails authn and the connection is closed,
+                    // but it is usually closed AFTER the handshake response returned
+                    logger.info("Connection will auto-close");
```

Comment:
This happens because the regular authentication is not configured. Opening remote cluster connection should have always failed because the remote_cluster_handshake is always rejected and openning connection does not return until remote_cluster_handshake is completed.

The same issue also affects `testSniffStrategyNoConnectionWhenAuthenticatorAlwaysFails`. Both of them have cause test failures.

For `testProxyStrategyCloseOnlyUnauthenticatedConnections`, even after configuring the regular authentication, it can still sometimes fail due this [assertion](https://github.com/elastic/elasticsearch/blob/25220057fd9722bda8de0818f3bd2a84504d3257/server/src/main/java/org/elasticsearch/transport/ProxyConnectionStrategy.java#L320) in ProxyConnectionStrategy. I am _not_ entirely sure why it happens. It could be a combination of issues like auto-connection retry, request sending retry and overlap of them. That said, I think we can drop this test because (1) it is a client side issue and (2) it is not a real situation. It is not possible for a legit client (a ES node) to have credentials changes in between requests for the same connection. So the client side issue will not actually happen. If the purpose is to test a malicious client gets disconnected, it is interfered by remoteClusterService's internal logic and not entirely functional because the client did get disconnected but it sometimes fails due to client side connection management which is not something we need to worry for a malicious client.

To put my comments in cocrete forms, here is a [diff](https://github.com/ywangd/elasticsearch/commit/ef7da3ce35b41746dc64a8e4591afdc54eaf9a8f) shows what I am suggesting. Please let me know what you think. Thanks!

Reply:
I don't think that using a `TransportInterceptor` in the test mock-up of the remote transport service, that does the in-line authn, is a proper way to test this PR.
The goal in this PR is to eventually close a connection upon finding the first invalid credentials. But unfortunately this is hard to test because of the connection strategies that obscure the disconnection.

The goal of `testProxyStrategyCloseOnlyUnauthenticatedConnections` is more to cover that the test infra fixture works correctly when authn is successful. The other tests implicitly assume that connection will not be established because authn is always reject, but it could be that there's something wrong in the mocks that really prevents it.

Reply:
I'll think some more about it, thanks for the feedback.

Reply:
We ended up adding the second in-line authn to the test fixture, and only test cases where all requests either pass (both) authn or none of them do. Then also add a test to ensure the connection is closed, outside the remote connection strategy infra.